File: packages/proximity-sensor/android/src/main/java/io/capawesome/capacitorjs/plugins/proximitysensor/ProximitySensorPlugin.java
package io.capawesome.capacitorjs.plugins.proximitysensor;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.getcapacitor.Logger;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import io.capawesome.capacitorjs.plugins.proximitysensor.classes.Measurement;
import io.capawesome.capacitorjs.plugins.proximitysensor.classes.interfaces.EmptyCallback;
import io.capawesome.capacitorjs.plugins.proximitysensor.classes.interfaces.NonEmptyResultCallback;
import io.capawesome.capacitorjs.plugins.proximitysensor.classes.interfaces.Result;
import io.capawesome.capacitorjs.plugins.proximitysensor.classes.results.GetMeasurementResult;
import io.capawesome.capacitorjs.plugins.proximitysensor.classes.results.IsAvailableResult;

@CapacitorPlugin(name = "ProximitySensor")
public class ProximitySensorPlugin extends Plugin {

    public static final String ERROR_UNKNOWN_ERROR = "An unknown error has occurred.";
    public static final String EVENT_MEASUREMENT = "measurement";
    public static final String TAG = "ProximitySensorPlugin";

    private final ProximitySensor implementation = new ProximitySensor(this);

    @PluginMethod
    public void getMeasurement(PluginCall call) {
        try {
            if (!implementation.isAvailable()) {
                rejectCallAsUnavailable(call);
                return;
            }

            NonEmptyResultCallback<GetMeasurementResult> callback = new NonEmptyResultCallback<>() {
                @Override
                public void success(@NonNull GetMeasurementResult result) {
                    resolveCall(call, result);
                }

                @Override
                public void error(Exception exception) {
                    rejectCall(call, exception);
                }
            };

            implementation.getMeasurement(callback);
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    @PluginMethod
    public void isAvailable(PluginCall call) {
        try {
            IsAvailableResult result = new IsAvailableResult(implementation.isAvailable());
            resolveCall(call, result);
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    public void notifyMeasurementEventListeners(@NonNull Measurement measurement) {
        notifyListeners(ProximitySensorPlugin.EVENT_MEASUREMENT, measurement.toJSObject());
    }

    @PluginMethod
    public void removeAllListeners(PluginCall call) {
        super.removeAllListeners(call);
        implementation.stopMeasurementUpdates(null);
    }

    @PluginMethod
    public void startMeasurementUpdates(PluginCall call) {
        try {
            if (!implementation.isAvailable()) {
                rejectCallAsUnavailable(call);
                return;
            }

            EmptyCallback callback = new EmptyCallback() {
                @Override
                public void success() {
                    resolveCall(call, null);
                }

                @Override
                public void error(Exception exception) {
                    rejectCall(call, exception);
                }
            };

            implementation.startMeasurementUpdates(callback);
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    @PluginMethod
    public void stopMeasurementUpdates(PluginCall call) {
        try {
            EmptyCallback callback = new EmptyCallback() {
                @Override
                public void success() {
                    resolveCall(call, null);
                }

                @Override
                public void error(Exception exception) {
                    rejectCall(call, exception);
                }
            };

            implementation.stopMeasurementUpdates(callback);
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    private void rejectCall(@NonNull PluginCall call, @NonNull Exception exception) {
        String message = exception.getMessage();
        if (message == null) {
            message = ERROR_UNKNOWN_ERROR;
        }
        Logger.error(TAG, message, exception);
        call.reject(message);
    }

    private void rejectCallAsUnavailable(@NonNull PluginCall call) {
        call.unavailable("This method is not available on this platform.");
    }

    private void resolveCall(@NonNull PluginCall call, @Nullable Result result) {
        if (result == null) {
            call.resolve();
        } else {
            call.resolve(result.toJSObject());
        }
    }
}

File: packages/proximity-sensor/ios/Plugin/ProximitySensorPlugin.swift
import Foundation
import Capacitor

@objc(ProximitySensorPlugin)
public class ProximitySensorPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "ProximitySensorPlugin"
    public let jsName = "ProximitySensor"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "getMeasurement", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "isAvailable", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "startMeasurementUpdates", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "stopMeasurementUpdates", returnType: CAPPluginReturnPromise)
    ]

    public static let measurementEvent = "measurement"

    private var implementation: ProximitySensor?

    override public func load() {
        self.implementation = ProximitySensor(plugin: self)
    }

    @objc func getMeasurement(_ call: CAPPluginCall) {
        guard implementation?.isAvailable() == true else {
            rejectCallAsUnavailable(call)
            return
        }

        implementation?.getMeasurement { measurement in
            self.resolveCall(call, measurement)
        }
    }

    @objc func isAvailable(_ call: CAPPluginCall) {
        let available = implementation?.isAvailable() ?? false
        resolveCall(call, IsAvailableResult(available: available))
    }

    @objc public func notifyMeasurementEventListeners(_ measurement: Measurement) {
        if let data = measurement.toJSObject() as? JSObject {
            notifyListeners(ProximitySensorPlugin.measurementEvent, data: data)
        }
    }

    @objc override public func removeAllListeners(_ call: CAPPluginCall) {
        super.removeAllListeners(call)
        implementation?.stopMeasurementUpdates()
    }

    @objc func startMeasurementUpdates(_ call: CAPPluginCall) {
        guard implementation?.isAvailable() == true else {
            rejectCallAsUnavailable(call)
            return
        }

        implementation?.startMeasurementUpdates {
            self.resolveCall(call)
        }
    }

    @objc func stopMeasurementUpdates(_ call: CAPPluginCall) {
        implementation?.stopMeasurementUpdates {
            self.resolveCall(call)
        }
    }

    private func rejectCallAsUnavailable(_ call: CAPPluginCall) {
        call.unavailable("This method is not available on this platform.")
    }

    private func resolveCall(_ call: CAPPluginCall) {
        call.resolve()
    }

    private func resolveCall(_ call: CAPPluginCall, _ result: Result?) {
        if let result = result?.toJSObject() as? JSObject {
            call.resolve(result)
        } else {
            call.resolve()
        }
    }
}
